src/tracing/esp32_diagnostic_trace/Diagnostics.h

```diff
@@ -0,0 +1,249 @@
+            {
+                char valueBuffer[kMaxStringValueSize + 1];
+                memcpy(valueBuffer, value_, kMaxStringValueSize);
+                valueBuffer[kMaxStringValueSize] = '\0';
+                ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::VALUE), valueBuffer));
+            }
+            else
+            {
+                ReturnErrorOnFailure(writer.PutString(chip::TLV::ContextTag(DiagTag::VALUE), value_));
+            }
+        }
+        else if constexpr (std::is_same_v<T, uint32_t>)
+        {
+            ReturnErrorOnFailure(writer.Put(chip::TLV::ContextTag(DiagTag::VALUE), value_));
+        }
+        else if constexpr (std::is_same_v<T, int32_t>)
+        {
+            ReturnErrorOnFailure(writer.Put(chip::TLV::ContextTag(DiagTag::VALUE), value_));
+        }
+        else
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+        ReturnErrorOnFailure(writer.EndContainer(DiagnosticOuterContainer));
+        ReturnErrorOnFailure(writer.Finalize());
+        ChipLogProgress(DeviceLayer, "Diagnostic Value written to storage successfully. label: %s\n", label_);
+        return CHIP_NO_ERROR;
+    }
+
+    CHIP_ERROR Decode(chip::TLV::TLVReader & reader) override
+    {
+        TLV::TLVType containerType;
+        ReturnErrorOnFailure(reader.EnterContainer(containerType));
+        // Read timestamp
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(DiagTag::TIMESTAMP)));
+        ReturnErrorOnFailure(reader.Get(timestamp_));
+
+        // Read label
+        ReturnErrorOnFailure(reader.Next(TLV::ContextTag(DiagTag::LABEL)));
+        uint32_t labelSize = reader.GetLength();
+        char labelBuffer[kMaxStringValueSize + 1];
+        ReturnErrorOnFailure(reader.GetString(labelBuffer, kMaxStringValueSize + 1));
+        memcpy(label_, labelBuffer, labelSize + 1);
+
+        // Read value
+        ReturnErrorOnFailure(reader.Next());
+        if constexpr (std::is_same_v<T, char *>)
+        {
+            uint32_t valueSize = reader.GetLength();
+            char valueBuffer[kMaxStringValueSize + 1];
+            ReturnErrorOnFailure(reader.GetString(valueBuffer, kMaxStringValueSize + 1));
+            memcpy(value_, valueBuffer, valueSize + 1);
+        }
+        else if constexpr (std::is_same_v<T, uint32_t>)
+        {
+            ReturnErrorOnFailure(reader.Get(value_));
+        }
+        else if constexpr (std::is_same_v<T, int32_t>)
+        {
+            ReturnErrorOnFailure(reader.Get(value_));
+        }
+        else
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+
+        ReturnErrorOnFailure(reader.ExitContainer(containerType));
+        return CHIP_NO_ERROR;
+    }
+
+    // Getters
+    char * GetLabel() const { return label_; }
+    T GetValue() const { return value_; }
+    uint32_t GetTimestamp() const { return timestamp_; }
+
+private:
+    char * label_ = nullptr;
+    T value_{};
+    uint32_t timestamp_ = 0;
```

Comment:
```suggestion
    const char * mLabel = nullptr;
    T mValue{};
    uint32_t mTimestamp = 0;
```

just to be more consistent with overall code base